Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -90,9 +90,8 @@ pb_bytes_array_t* Serializer::EncodeBytes(const std::vector<uint8_t>& bytes) {
   return ByteString(bytes).release();
 }
 
-std::vector<uint8_t> Serializer::DecodeBytes(const pb_bytes_array_t* bytes) {
-  if (bytes == nullptr) return {};
-  return std::vector<uint8_t>(bytes->bytes, bytes->bytes + bytes->size);
+ByteString Serializer::DecodeBytes(const pb_bytes_array_t* bytes) {
+  return ByteString(bytes);
```

Comment:
You've removed the `if (bytes==nullptr) return {};`. As it currently stands, this method will segfault if given a nullptr.

I don't think it was ever possible for nullptr to be passed. Even if the proto message is corrupted, I think you could get garbage values (incl 0s obviously) for the contents, but I think nanopb would always return *something*. I'm not 100% certain of that though.

I'd recommend either:
a) restoring the original behaviour.
b) hard_assert != nullptr. Either here or in the ByteString ctor (or both) and then adjusting the doc string in `serializer.h:118` to change this `If the input pointer is null, then this method will return an empty vector.` to `@pre bytes != nullptr` or similar.

Reply:
Ugh. `ByteString` previously sort of tolerated its bytes being null, though this was only the default constructed/moved-from state but not really intended for the empty case. Over time you've been pushing me to assert that the member `bytes_` is not `nullptr`. This particular case has me thinking that the approach is flawed.

In particular, if `ByteString` has to deal with instances of itself being valid that have no bytes, there seems little reason to force every caller to deal with this ambiguity.

Looking back at the docs, proto3 doesn't distinguish between empty and missing string/bytes values. To me, this means that our attempting to do so here is making a distinction that shouldn't really matter. Indeed I've added a test (`EncodesNullBlobs`) that shows that this behavior holds.

So, concretely, changes I've made in response to this feedback:
  * `ByteString` now explicitly avoids allocation in the empty case. `ByteString()` and `ByteString("")` are now indistinguishable.
  * Default, empty, and moved-from `ByteString`s all have the same representation.
  * Callers no longer have to handle null `pb_bytes_array_t*` values. This code is now correct as written.
  * I've stripped all the `HARD_ASSERT(bytes_ != nullptr)` out of the implementation.

Reply:
OTOH, consumers now have to handle null return values from ByteString.get() and .release(). (Though curiously not ToNSData() which will allocate a 0 length buffer.)

You could:
a) Change the default/etc ctor to always create a zero-length buffer. (Allowing input of nullptr and then convert to a zero-length buffer.) This would allow you to remove the special casing of nullptr as all methods could then assume a non-null `bytes_*`.
b) Change .get() and .release() to return non-null, even when bytes_ == nullptr. (get() could return a pointer to some static (constexpr?) value. release() would... I guess have to alloc. But I suppose that's what ToNSData() does now.)
c) Just make the callers deal with nullptrs. For consistency, consider altering ToNSData to also return nullptr when bytes_ == nullptr.

proto3 doesn't distinguish between empy/missing bytes on input, but when you retrieve the bytes, I believe it returns an empty byte array (rather than null) in both cases. If we're going to mimic it's behaviour, then this would argue for (a) or (b).

But at this point, we're bikesheding. What you've got is fine. Only requirement from me is to at least consider the above before hitting the submit button. :)